### PR TITLE
fix(js): Avoid dropping parentheses when autofixing

### LINKED ISSUE #6233: Autofix mangles TypeScript expression containing parentheses
**Describe the bug**
I found a case where the autofix functionality incorrectly transforms a TypeScript expression containing parentheses by eating an opening parenthesis.

**To Reproduce**
Steps to reproduce the behavior, ideally a link to https://semgrep.dev: https://semgrep.dev/s/Aq24

**Expected behavior**

The transformed expression should be

```typescript
function test(obj, first, second, val) {
  obj[first][second] = val
}

test(m, t, (e as SomeType).some_key, e as SomeType);
```

But instead it is

```typescript
function test(obj, first, second, val) {
  obj[first][second] = val
}

test(m, t, e as SomeType).some_key, e as SomeType);
        //^ Note the missing opening parenthesis.
```

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [x] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

### PR DESCRIPTION
**Pull request by nmote**
This is equivalent to #7437, but for JS. Now we include parentheses in the range for expressions, so we don't end up in weird situations where the location of an expression spans one parenthesis but not its match.

Fixes #6233

Probably fixes a couple other issues too (will check after).

Test plan:

* Automated tests
* Manual test of the case given in #6233
* Put the Menhir parser first in `Parse_target2.ml` then manual test of https://semgrep.dev/s/EAYw.

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: languages/javascript/generic/js_to_generic.ml [hunk 1]
```diff
@@ -223,60 +223,65 @@ and literal x : G.literal =
...
+      G.Cast (v3, v2, v1) |> G.e
   | ExprTodo (v1, v2) ->
       let v2 = list expr v2 in
-      G.OtherExpr (v1, v2 |> Common.map (fun e -> G.E e))
-  | L x -> G.L (literal x)
+      G.OtherExpr (v1, v2 |> Common.map (fun e -> G.E e)) |> G.e
+  | ParenExpr (l, e, r) ->
+      let e = expr e in
+      H.set_e_range l r e;
+      e
```
Comment (emjin) [inline languages/javascript/generic/js_to_generic.ml:246]: Out of curiosity, why did you choose this rather than translating to an AST_generic.ParenExpr?

**Review by emjin — APPROVED** [review]

File: languages/javascript/generic/js_to_generic.ml (hunk 1 above)
Comment (nmote) [reply languages/javascript/generic/js_to_generic.ml:246]: See #7437 for a more detailed explanation, but basically when I tried that with Python it broke a whole bunch of stuff

Comment (nmote) [reply languages/javascript/generic/js_to_generic.ml:246]: We actually should probably delete ParenExpr from the AST and write to `e_range` instead for the one spot it's actually used

**Comment by nmote** [comment]
Also fixes #1917

**Comment by nmote** [comment]
Also fixes the autofix problems with #2051 (after a tweak to the metavariable-regex to allow the parens that are now part of the expression).

**Pull request merged**